pyro/infer/mcmc/api.py

```diff
@@ -0,0 +1,332 @@
+
+    return _add_logging
+
+
+class _UnarySampler(object):
+    """
+    Single process runner class optimized for the case `num_chains=1`.
+    """
+
+    def __init__(self, kernel, num_samples, warmup_steps, disable_progbar, initial_params=None, hook=None):
+        self.kernel = kernel
+        if initial_params is not None:
+            self.kernel.initial_params = initial_params
+        self.warmup_steps = warmup_steps
+        self.num_samples = num_samples
+        self.logger = None
+        self.disable_progbar = disable_progbar
+        self.hook = hook
+        super(_UnarySampler, self).__init__()
+
+    def run(self, *args, **kwargs):
+        logger = logging.getLogger("pyro.infer.mcmc")
+        progress_bar = ProgressBar(self.warmup_steps, self.num_samples, disable=self.disable_progbar)
+        logger = initialize_logger(logger, "", progress_bar)
+        hook_w_logging = _add_logging_hook(logger, progress_bar, self.hook)
+        for sample in _gen_samples(self.kernel, self.warmup_steps, self.num_samples, hook_w_logging,
+                                   *args, **kwargs):
+            yield sample, 0  # sample, chain_id (default=0)
+        progress_bar.close()
+
+
+class _MultiSampler(object):
+    """
+    Parallel runner class for running MCMC chains in parallel. This uses the
+    `torch.multiprocessing` module (itself a light wrapper over the python
+    `multiprocessing` module) to spin up parallel workers.
+    """
+    def __init__(self, kernel, num_samples, warmup_steps, num_chains, mp_context,
+                 disable_progbar, initial_params=None, hook=None):
+        self.kernel = kernel
+        self.warmup_steps = warmup_steps
+        self.num_chains = num_chains
+        self.hook = hook
+        self.workers = []
+        self.ctx = mp
+        if mp_context:
+            if six.PY2:
+                raise ValueError("multiprocessing.get_context() is "
+                                 "not supported in Python 2.")
+            self.ctx = mp.get_context(mp_context)
+        self.result_queue = self.ctx.Queue()
+        self.log_queue = self.ctx.Queue()
+        self.logger = initialize_logger(logging.getLogger("pyro.infer.mcmc"),
+                                        "MAIN", log_queue=self.log_queue)
+        self.num_samples = num_samples
+        self.initial_params = initial_params
+        self.log_thread = threading.Thread(target=logger_thread,
+                                           args=(self.log_queue, self.warmup_steps, self.num_samples,
+                                                 self.num_chains, disable_progbar))
+        self.log_thread.daemon = True
+        self.log_thread.start()
+        self.events = [self.ctx.Event() for _ in range(num_chains)]
+
+    def init_workers(self, *args, **kwargs):
+        self.workers = []
+        for i in range(self.num_chains):
+            init_params = {k: v[i] for k, v in self.initial_params.items()}
```

Comment:
I think that this won't work if `initial_params=None`. So for my understanding, here are what is working and what not, which we can address in follow up PRs:
+ with `potential_fn`:
  - need an explicit kwargs `transforms` either in HMC or MCMC.
  - need an explicit `inital_params`, in both 1 chain and multi-chain
+ with `model`:
  - need an explicit `initial_params` for multi-chain

I think we can resolve part of them with
+ with `potential_fn`, use `transforms={}` if it is not provided
+ with `potential_fn`, fix HMC setup so that if  `initial_params` is None, then sample `initial_params` (maybe in (-2, 2)) interval
+ for both case `potential_fn` or `model`, change this code to
```
init_params = {k: v[i] for k, v in self.initial_params.items()} if self.initial_params is not None else None
```

Reply:
All excellent points. I was planning to address this with the next PR that gets the baseball example working, but I think these are small changes that we might as well fix right now. I'll address them shortly.

Reply:
Actually, (2) isn't possible. We need to explicitly provide `initial_params` with `potential_fn` since we don't know the data structure / site shapes etc. that the `potential_fn` takes which is additional information that the user needs to provide.

Reply:
Oppss, you are right. It is good to add warning as you did. +1